### PR TITLE
SONARSLANG-238 [Scala] Validate Slang AST completeness

### PR DESCRIPTION
**Pull request by alban-auzeill**

### CODE REVIEW

File: slang-plugin/src/test/java/org/sonarsource/slang/plugin/SlangSensorTest.java [hunk 1]
```diff
@@ -60,6 +62,12 @@
 
 public class SlangSensorTest extends AbstractSensorTest {
 
+  @Before
+  public void setup() {
+    super.setup();
+    context.setSettings(new MapSettings().setProperty("sonar.slang.converter.validation", "log"));
+  }
+
```
Comment (pynicolas) [inline slang-plugin/src/test/java/org/sonarsource/slang/plugin/SlangSensorTest.java]: Do we need that change?

File: sonar-scala-plugin/src/test/java/org/sonarsource/scala/plugin/ScalaSensorTest.java [hunk 2]
```diff
@@ -33,6 +34,12 @@
 
 public class ScalaSensorTest extends AbstractSensorTest {
 
+  @Before
+  public void setup() {
+    super.setup();
+    context.setSettings(new MapSettings().setProperty("sonar.slang.converter.validation", "throw"));
+  }
+
```
Comment (pynicolas) [inline sonar-scala-plugin/src/test/java/org/sonarsource/scala/plugin/ScalaSensorTest.java]: Do we need that change?

File: its/plugin/src/test/java/org/sonarsource/slang/TestBase.java [hunk 3]
```diff
@@ -62,6 +62,7 @@ protected SonarScanner getSonarScanner(String directoryToScan, String languageKe
       .setProjectDir(new File(directoryToScan, languageKey))
       .setProjectKey(PROJECT_KEY)
       .setProjectName(PROJECT_KEY)
+      .setProperty("sonar.slang.converter.validation", "log")
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply its/plugin/src/test/java/org/sonarsource/slang/TestBase.java]: I will remove it

File: slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+import org.sonarsource.slang.api.TextRange;
+import org.sonarsource.slang.api.Token;
+import org.sonarsource.slang.api.TopLevelTree;
+import org.sonarsource.slang.api.Tree;
+import org.sonarsource.slang.api.TreeMetaData;
+import org.sonarsource.slang.impl.LiteralTreeImpl;
+import org.sonarsource.slang.impl.NativeTreeImpl;
+import org.sonarsource.slang.impl.TextPointerImpl;
+
+public class ASTConverterValidation implements ASTConverter {
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java]: I will move it

File: slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+  @Override
+  public void terminate() {
+    List<String> errors = errors();
+    if (!errors.isEmpty()) {
+      LOG.error("AST Converter Validation:\n  [AST ERROR] " + String.join("\n  [AST ERROR] ", errors));
+    }
+    wrapped.terminate();
+  }
+
+  public ValidationMode mode() {
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java]: I will change to visible for testing

File: slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    return mode;
+  }
+
+  public List<String> errors() {
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java]: I will change to visible for testing

File: slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    if (tree instanceof IdentifierTree) {
+      unexpectedTokens = tokens.stream()
+        .filter(token -> token.type() == Token.Type.KEYWORD || token.type() == Token.Type.STRING_LITERAL)
+        .collect(Collectors.toList());
+    } else {
+      unexpectedTokens = tokens.stream()
+        .filter(token -> !(token.type() == Token.Type.KEYWORD || token.type() == Token.Type.STRING_LITERAL))
+        .filter(token -> !PUNCTUATOR_PATTERN.matcher(token.text()).matches())
+        .collect(Collectors.toList());
+    }
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java]: yes 'true' and 'false'

File: slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+      return;
+    }
+    List<Token> unexpectedTokens;
+    if (tree instanceof IdentifierTree) {
+      unexpectedTokens = tokens.stream()
+        .filter(token -> token.type() == Token.Type.KEYWORD || token.type() == Token.Type.STRING_LITERAL)
+        .collect(Collectors.toList());
+    } else {
+      unexpectedTokens = tokens.stream()
+        .filter(token -> !(token.type() == Token.Type.KEYWORD || token.type() == Token.Type.STRING_LITERAL))
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply slang-api/src/main/java/org/sonarsource/slang/utils/ASTConverterValidation.java]: I will remove Token.Type.STRING_LITERA

File: slang-api/src/test/java/org/sonarsource/slang/utils/ASTConverterValidationTest.java [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+      }
+
+      @Override
+      public List<Token> tokens() {
+        return tokens;
+      }
+
+      @Override
+      public Set<Integer> linesOfCode() {
+        return tokens.stream()
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply slang-api/src/test/java/org/sonarsource/slang/utils/ASTConverterValidationTest.java]: I will throw unsupported ....

**Pull request merged**